### PR TITLE
Force deabstraction of functions that take or return Tensors if referred in code.

### PR DESCRIPTION
**Pull request by bgogul**
This is a somewhat hacky patch to deal with [SR-8589](https://bugs.swift.org/browse/SR-8589) related to the Jupyter tutoral. In the REPL mode (both in the in-build swift and lldb -r), the SIL module for a REPL expr is thrown away after lowering to LLVM IR. Consequently, when processing subsequent lines, TFDeabstraction does not see the body of some functions being invoked, which causes some functions not to be partitioned. 

This PR performs a second round of deabstraction (and subsequent partitioning) of functions taking and returning TensorValues if they are still referred in the code after the first round. 

BTW, this is not only relevant to Jupyter tutorial. We will encounter this if we allow linking to other swift-tensorflow binaries. We will need to think about how to deal with such cases in general.

@mhong, when I simply removed the condition that checks if a function operates on tensors, some tests started failing as attributes were no longer const-evaluable. That is why I am sending this instead of what we discussed.

### CODE REVIEW

**Review by lattner — APPROVED** [review]
Agreed that this is not optimal, but I think it should work in the near term.  Please file a bug to track improving this.  A sketch:

1) Make our existing pass over functions, deabstracting as we go.

2) Do a second pass where we deabstract all functions, regardless of whether they have tensor args or results.  Many would be dead, but the indirectly used ones would now work.

**Review by rxwei — APPROVED** [review]
The fix makes sense, unblocking lots of cross-module cases. 

There is a more general problem though: whenever there’s a call to an opaque function which calls a protocol method (e.g. Equatable.==) passing tensor arguments whose implementation of this protocol is inlinable, execution will fail. This is because Tensor’s protocol implementation is never getting partitioned.

To resolve the bigger problem later down the road, partitioning generic functions should be supported and will require a major architectural change.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 1]
```diff
@@ -2522,13 +2524,41 @@ void TFDeabstractionPass::run() {
...
     // passes are just function passes that are invoked by this one.  Until
     // we are ready for that, let them run later in the pipeline after the
     // other optimization and cleanup passes.
   }
+
+  // Deabstract stragglers that were left out in the previous iteration. These
+  // are functions that are *still* referred to in the code and operate on
+  // tensor values, but have not been partitioned. It can happen in the
+  // following case, for instance, where `foo` is an external function that
+  // takes a for which we do not have the body:
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: what is "takes a for"?

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 2]
```diff
@@ -2522,13 +2524,41 @@ void TFDeabstractionPass::run() {
...
+  //   main() {
+  //     foo() { $0 -= 0.1 * $1 }
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: can we turn something like this into a unit test?

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 3]
```diff
@@ -2522,13 +2524,41 @@ void TFDeabstractionPass::run() {
...
+  //
+  // In the common case where we have the body of foo(), it gets inlined, and
+  // therefore, the closure gets inlined too. However, if the body of foo() is
+  // not available, the closure never gets inlined. To ensure that the call to
+  // the closure within foo() works correctly, we will have to deabstract and
+  // partition the closure.
+  //
+  // (Note the body of a function may be missing when we are linking against a
+  // library or in the REPL context where the function was defined on a
+  // different line.)
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: nit: "different line" might not be the right description.

@marcrasi showed me earlier that you can copy/paste multi-line text into REPL, and that'll be executed as one "cell".

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 4]
```diff
@@ -2522,13 +2524,41 @@ void TFDeabstractionPass::run() {
...
+  for (auto &fn : *module) {
+    // Skip if it is already partitioned, or if it was ignored only because it
+    // operated on tensor values.
+    if (partitionedFunctions.count(&fn) > 0 ||
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: can you add a comment block motivating this two-round / two-pass design?

IIUC, in your example above, if we just use `tfc.shouldBePartitioned(&fn, /*forceTFFunctions=*/true)` in the first round, and make it the only round, we'll be able to partition the private closure, so it achieves the same correctness.

the two-round design is an optimization to minimize the set of functions to partition (would be nice to have a unit test that shows its effects -- have a function with no caller left after the first round finishes, and confirm it's not partitioned in the second round).

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 5]
```diff
@@ -631,7 +631,8 @@ tf::createTensorToInt1Inst(SILValue value, SILBuilder &builder,
 /// example) for inlined functions that take and return tensors, since we
 /// know that they are either unreachable or will be inlined into any
 /// clients that use them.
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: consider removing this comment block as it's now out of date (the new param is not documented).

(i'd also be fine to keep it in sync with the corresponding comment block in the header file.)

**Review by mhong — APPROVED** [review]
Having this as a short-term patch sounds good, but please add some unit test.

To avoid breaking unit tests, one option is to flag-control the value of `forceTFFunctions`, and only turn it on in REPL mode. This will buy us time to rewrite the relevant unit tests -- if we don't want a function to be partitioned (e.g. if the func takes a string/int/float param, and that param is used in a tfop attr), we can mark it " inline always". Once we finish this on all tests, we can then retire that flag.

**Comment by mhong** [comment]
@rxwei To help me understand this larger problem, can you provide a concrete example? 

(It might be useful to create a bug to track that.)

**Comment by rxwei** [comment]
A common, concrete example is to try to use `expectEqual()` two `Tensor`s. `expectEqual` is opaque across module, i.e. non-inlinable.

This is not a small bug, but a large hole in the architecture in that there's no way to even reject it from the compiler. These issues are not important in the initial GPE because the library ecosystem using `Tensor`s and generics are not there yet. 

I'll definitely file a bug soon. I was also planning to write a detailed document describing the issue, but it's not the best time until we hit the tutorial goal.

**Comment by bgogul** [comment]
@lattner , this PR already implements it in two passes as you outlined. Did I misunderstand your comment? 

> Agreed that this is not optimal, but I think it should work in the near term. Please file a bug to track improving this. A sketch:
> 
> Make our existing pass over functions, deabstracting as we go.
> 
> Do a second pass where we deabstract all functions, regardless of whether they have tensor args or results. Many would be dead, but the indirectly used ones would now work.

**Comment by mhong** [comment]
@bgogul, I believe the difference is: If we change the example from
```swift
main() {
  foo() { $0 -= 0.1 * $1 }
}
```

to 
```swift
main() {
  foo() { #tfop("SomeOp", ...) }
}
```

This patch probably won't work. That is, if the private closure does not take or return tensors.

Another issue to take care (both in Chris' sketch and this PR) is that if the private closure takes a string/int/float param to be used in a tfop attr, as in:
```swift
main() {
  foo() { (s: String, i: Int) -> () in #tfop("SomeOp", intAttr: i, sttrAttr: s) }
}
```

The private closure cannot be partitioned due to const tfop attr requirement (as you observed). In that case it'd be useful to issue a good diagnostic message, asking user to convert the closure to a function, so that it can be annotated with something like "inline always", and as such compiler won't partition it.

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 6]
```diff
@@ -704,8 +705,16 @@ bool TensorFunctionClassifier::shouldBePartitioned(SILFunction *fn) {
...
   // that it will be inlined away by deabstraction, and we don't need to touch
   // it.
-  if (containsTensorFlowValue(fn->getLoweredFunctionType()))
-    return false;
+  if (containsTensorFlowValue(fn->getLoweredFunctionType())) {
+    if (forceTFFunctions) {
+      // Return true if it is referenced somewhere.
+      // TODO: There might be a better check other than fn->getRefCount() > 0.
+      // See the clean up code in TFDeabstraction::inlineCalls() function.
+      return (fn->getRefCount() > 0);
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFUtilities.h [hunk 7]
```diff
@@ -277,7 +277,11 @@ class TensorFunctionClassifier {
   /// example) for inlined functions that take and return tensors, since we
   /// know that they are either unreachable or will be inlined into any
   /// clients that use them.
-  bool shouldBePartitioned(SILFunction *fn);
+  ///
+  /// If the flag forceTFFunctions is true, it forces partitioning of functions
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFUtilities.h]: Done.

File: lib/SILOptimizer/Mandatory/TFUtilities.h [hunk 8]
```diff
@@ -277,7 +277,11 @@ class TensorFunctionClassifier {
...
+  /// that operate on Tensor values irrespective of whether they are inlinable,
+  /// private, etc.
+  bool shouldBePartitioned(SILFunction *fn, bool forceTFFunctions = false);
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFUtilities.h]: Done.

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 5 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Removed the block comment altogether.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 4 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Added a few lines motivating the two round design.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 3 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Calling it REPL line now. Not sure if that is the right terminology.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 2 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Added a unit test in deabstraction_finished.swift

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 1 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: typo. Fixed the comment.

**Comment by bgogul** [comment]
Added unit tests and addressed comments. PTAL.

**Review by mhong — APPROVED** [review]
Nice!

**Pull request merged**